pommerman/network/ion_server/network.py

```diff
@@ -0,0 +1,205 @@
+                QUEUE_MAIN.send([CONCURRENTLY_LOOKING, len(PLAYER_WS)])
+            elif (queue_msg[0] == spc.update_cc.value):
+                CONCURRENTLY_LOOKING = queue_msg[1]
+        if not QUEUE_SUBPROC.empty():
+            queue_msg = QUEUE_SUBPROC.get()
+            MATCH_PROCESS[queue_msg[2]] = {
+                "p": queue_msg[0],
+                "pl": queue_msg[1],
+                "u": queue_msg[2],
+                "f": False,
+                "d": False
+            }
+            for i in queue_msg[1]:
+                if i in PLAYER_WS:
+                    await PLAYER_WS[i]["ws"].send(
+                        dumps({
+                            "i": nic.match_start.value,
+                            "u": queue_msg[2]
+                        }))
+        for key in list(MATCH_PROCESS.keys()):
+            value = MATCH_PROCESS[key]
+            if (value["p"].poll() and not i["f"]):
+                pipe_msg = value["p"].recv()
+                if (pipe_msg[0] == spc.match_next.value):
+                    value["f"] = not i["f"]
+                    value["act"] = [0, 0, 0, 0]
+                    value["recv"] = [False, False, False, False]
+                    value["time"] = time()
+                    value["id"] = pipe_msg[1]
+                    for x, y in enumerate(value["pl"]):
+                        if y in list(PLAYER_WS.keys()):
+                            try:
+                                await PLAYER_WS[y]["ws"].send(pipe_msg[2][x])
+                            except:
+                                pass
+                        elif y not in PLAYER_WS:
+                            i["act"][x] = 5
+                if (pipe_msg[0] == spc.match_end.value):
+                    value["d"] = True
+                    for x, y in enumerate(i["pl"]):
+                        if y in PLAYER_WS:
+                            await PLAYER_WS[y]["ws"].send(
+                                dumps({
+                                    "i": nic.match_end.value,
+                                    "r": pipe_msg[1][x]
+                                }))
+            if value["f"]:
+                if (value["time"] + STOP_TIMEOUT < time()) or all(
+                        value["recv"]) is True:
+                    value["p"].send(i["act"])
+                    value["f"] = not i["f"]
+            if value["d"]:
+                value["p"].send("END")
+                del MATCH_PROCESS[key]
+        THREAD_LOCK = False
+        sleep(
+            0.0001
+        )  # Sleep for the a extremely small amount of time so the other thread can take the lock
+
+
+def thread(queue_main, queue_subproc, port, max_players, mode,
+           stop_timeout):
+    """Creates a network thread"""
+    global MAX_PLAYERS, QUEUE_MAIN, QUEUE_SUBPROC, MODE, STOP_TIMEOUT
+
+    MAX_PLAYERS = max_players
+    QUEUE_MAIN = queue_main
+    QUEUE_SUBPROC = queue_subproc
+    MODE = mode
+    STOP_TIMEOUT = stop_timeout
```

Comment:
Don't instantiate global variables in a local function. For one thing, this makes code evaluation absurdly hard. The first time that I came across any of these was half way through ```program_loop```.

Declare everything up front at the top of the module. Then we can evaluate what's necessary, add comments, and understand the flow much easier.

Reply:
Done 👍